File: server/src/main/java/org/apache/iotdb/db/query/udf/api/customizer/parameter/UDFParameterValidator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iotdb.db.query.udf.api.customizer.parameter;

import org.apache.iotdb.db.exception.metadata.MetadataException;
import org.apache.iotdb.db.query.udf.api.exception.UDFAttributeNotProvidedException;
import org.apache.iotdb.db.query.udf.api.exception.UDFException;
import org.apache.iotdb.db.query.udf.api.exception.UDFInputSeriesDataTypeNotValidException;
import org.apache.iotdb.db.query.udf.api.exception.UDFInputSeriesIndexNotValidException;
import org.apache.iotdb.db.query.udf.api.exception.UDFInputSeriesNumberNotValidException;
import org.apache.iotdb.db.query.udf.api.exception.UDFParameterNotValidException;
import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;

public class UDFParameterValidator {

  private final UDFParameters parameters;

  public UDFParameterValidator(UDFParameters parameters) {
    this.parameters = parameters;
  }

  public UDFParameters getParameters() {
    return parameters;
  }

  /**
   * Validates whether the attributes entered by the user contain an attribute whose key is
   * attributeKey.
   *
   * @param attributeKey key of the attribute
   * @throws UDFAttributeNotProvidedException if the attribute is not provided
   */
  public UDFParameterValidator validateRequiredAttribute(String attributeKey)
      throws UDFAttributeNotProvidedException {
    if (!parameters.hasAttribute(attributeKey)) {
      throw new UDFAttributeNotProvidedException(attributeKey);
    }
    return this;
  }

  /**
   * Validates whether the data type of the input series at the specified column is as expected.
   *
   * @param index index of the specified column
   * @param expectedDataType the expected data type
   * @throws UDFInputSeriesIndexNotValidException if the index of the specified column is out of
   *     bound
   * @throws UDFInputSeriesDataTypeNotValidException if the data type of the input series at the
   *     specified column is not as expected
   */
  public UDFParameterValidator validateInputSeriesDataType(int index, TSDataType expectedDataType)
      throws UDFException {
    validateInputSeriesIndex(index);

    TSDataType actualDataType;
    try {
      actualDataType = parameters.getDataType(index);
    } catch (MetadataException e) {
      throw new UDFException("error occurred when getting the data type of the input series");
    }

    if (!expectedDataType.equals(actualDataType)) {
      throw new UDFInputSeriesDataTypeNotValidException(index, actualDataType, expectedDataType);
    }
    return this;
  }

  /**
   * Validates whether the data type of the input series at the specified column is as expected.
   *
   * @param index index of the specified column
   * @param expectedDataTypes the expected data types
   * @throws UDFInputSeriesIndexNotValidException if the index of the specified column is out of
   *     bound
   * @throws UDFInputSeriesDataTypeNotValidException if the data type of the input series at the
   *     specified column is not as expected
   */
  public UDFParameterValidator validateInputSeriesDataType(
      int index, TSDataType... expectedDataTypes) throws UDFException {
    validateInputSeriesIndex(index);

    TSDataType actualDataType;
    try {
      actualDataType = parameters.getDataType(index);
    } catch (MetadataException e) {
      throw new UDFException("error occurred when getting the data type of the input series");
    }

    for (TSDataType expectedDataType : expectedDataTypes) {
      if (expectedDataType.equals(actualDataType)) {
        return this;
      }
    }

    throw new UDFInputSeriesDataTypeNotValidException(index, actualDataType, expectedDataTypes);
  }

  /**
   * Validates whether the number of the input series is as expected.
   *
   * @param expectedSeriesNumber the expected number of the input series
   * @throws UDFInputSeriesNumberNotValidException if the number of the input series is not as
   *     expected
   */
  public UDFParameterValidator validateInputSeriesNumber(int expectedSeriesNumber)
      throws UDFInputSeriesNumberNotValidException {
    int actualSeriesNumber = parameters.getPaths().size();
    if (actualSeriesNumber != expectedSeriesNumber) {
      throw new UDFInputSeriesNumberNotValidException(actualSeriesNumber, expectedSeriesNumber);
    }
    return this;
  }

  /**
   * Validates whether the number of the input series is as expected.
   *
   * @param expectedSeriesNumberLowerBound the number of the input series must be greater than or
   *     equal to the expectedSeriesNumberLowerBound
   * @param expectedSeriesNumberUpperBound the number of the input series must be less than or equal
   *     to the expectedSeriesNumberUpperBound
   * @throws UDFInputSeriesNumberNotValidException if the number of the input series is not as
   *     expected
   */
  public UDFParameterValidator validateInputSeriesNumber(
      int expectedSeriesNumberLowerBound, int expectedSeriesNumberUpperBound)
      throws UDFInputSeriesNumberNotValidException {
    int actualSeriesNumber = parameters.getPaths().size();
    if (actualSeriesNumber < expectedSeriesNumberLowerBound
        || expectedSeriesNumberUpperBound < actualSeriesNumber) {
      throw new UDFInputSeriesNumberNotValidException(
          actualSeriesNumber, expectedSeriesNumberLowerBound, expectedSeriesNumberUpperBound);
    }
    return this;
  }

  /**
   * Validates the input parameters according to the validation rule given by the user.
   *
   * @param validationRule the validation rule, which can be a lambda expression
   * @param messageToThrow the message to throw when the given argument is not valid
   * @param argument the given argument
   * @throws UDFParameterNotValidException if the given argument is not valid
   */
  public UDFParameterValidator validate(
      SingleObjectValidationRule validationRule, String messageToThrow, Object argument)
      throws UDFParameterNotValidException {
    if (!validationRule.validate(argument)) {
      throw new UDFParameterNotValidException(messageToThrow);
    }
    return this;
  }

  public interface SingleObjectValidationRule {

    boolean validate(Object arg);
  }

  /**
   * Validates the input parameters according to the validation rule given by the user.
   *
   * @param validationRule the validation rule, which can be a lambda expression
   * @param messageToThrow the message to throw when the given arguments are not valid
   * @param arguments the given arguments
   * @throws UDFParameterNotValidException if the given arguments are not valid
   */
  public UDFParameterValidator validate(
      MultipleObjectsValidationRule validationRule, String messageToThrow, Object... arguments)
      throws UDFParameterNotValidException {
    if (!validationRule.validate(arguments)) {
      throw new UDFParameterNotValidException(messageToThrow);
    }
    return this;
  }

  public interface MultipleObjectsValidationRule {

    boolean validate(Object... args);
  }

  /**
   * Validates whether the index of the specified column is out of bound. bound: [0,
   * parameters.getPaths().size())
   *
   * @param index the index of the specified column
   * @throws UDFInputSeriesIndexNotValidException if the index of the specified column is out of
   *     bound
   */
  private void validateInputSeriesIndex(int index) throws UDFInputSeriesIndexNotValidException {
    int actualSeriesNumber = parameters.getPaths().size();
    if (index < 0 || actualSeriesNumber <= index) {
      throw new UDFInputSeriesIndexNotValidException(index, actualSeriesNumber);
    }
  }
}

File: udf-py/udf_api/customizer/parameter/udf_parameter_validator.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import List

from udf_api.customizer.parameter.udf_parameters import UDFParameters
from udf_api.exception.udf_attribute_not_provided_exception import (
    UDFAttributeNotProvidedException,
)
from udf_api.exception.udf_input_series_data_type_not_valid_exception import (
    UDFInputSeriesDataTypeNotValidException,
)
from udf_api.exception.udf_input_series_index_not_valid_exception import (
    UDFInputSeriesIndexNotValidException,
)
from udf_api.exception.udf_input_series_number_not_valid_exception import (
    UDFInputSeriesNumberNotValidException,
)
from udf_api.exception.udf_parameter_not_valid_exception import (
    UDFParameterNotValidException,
)
from udf_api.type.type import Type


class UDFParameterValidator:
    __parameters: UDFParameters

    def __init__(self, parameters: UDFParameters):
        self.__parameters = parameters

    def get_parameters(self) -> UDFParameters:
        return self.__parameters

    def validate_required_attribute(self, attribute_key: str):
        """
        Validates whether the attributes entered by the user contain an attribute whose key is
        attributeKey.

        :param attribute_key: key of the attribute
        :raise UDFAttributeNotProvidedException: if the attribute is not provided
        """
        if not self.__parameters.has_attribute(attribute_key):
            raise UDFAttributeNotProvidedException(attribute_key)

        return self

    def validate_input_series_data_type(
        self, index: int, expected_data_types: List[Type]
    ):
        """
        Validates whether the data type of the input series at the specified column is as expected.

        :param index: index of the specified column
        :param expected_data_types: the expected data types
        :raise UDFInputSeriesIndexNotValidException: if the index of the specified column is out of bound
        :raise UDFInputSeriesDataTypeNotValidException: if the data type of the input series at the specified column is
        not as expected
        """
        self.validate_input_series_index(index)

        actual_data_type = self.__parameters.get_child_expression_data_types()[index]

        for expected_data_type in expected_data_types:
            if expected_data_type == actual_data_type:
                return self

        raise UDFInputSeriesDataTypeNotValidException(
            index, actual_data_type, expected_data_types
        )

    def validate_input_series_number(
        self,
        expected_series_number_lower_bound: int,
        expected_series_number_upper_bound,
    ):
        """
        Validates whether the number of the input series is as expected.

        :param expected_series_number_lower_bound: the number of the input series must be greater than or equal to the
        expected_series_number_lower_bound
        :param expected_series_number_upper_bound: the number of the input series must be less than or equal to the
        expected_series_number_upper_bound
        :raise UDFInputSeriesNumberNotValidException: if the number of the input series is not as expected
        """
        actual_series_number = self.__parameters.get_child_expressions_count()
        if (
            actual_series_number < expected_series_number_lower_bound
            or expected_series_number_upper_bound < actual_series_number
        ):
            raise UDFInputSeriesNumberNotValidException(
                actual_series_number,
                expected_series_number_lower_bound,
                expected_series_number_upper_bound,
            )

        return self

    def validate(self, message_to_raise: str, validation_rule, **kwargs):
        """
        Validates the input parameters according to the validation rule given by the user.

        :param message_to_raise: the message to throw when the given arguments are not valid
        :param validation_rule: the validation rule, which can be a lambda expression
        :param kwargs: the given arguments
        :raise UDFParameterNotValidException: if the given arguments are not valid
        """
        if not validation_rule(kwargs):
            raise UDFParameterNotValidException(message_to_raise)

        return self

    def validate_input_series_index(self, index: int):
        """
        Validates whether the index of the specified column is out of bound. bound: [0,
        self.__parameters.get_child_expressions_count())

        :param index: the index of the specified column
        :raise UDFInputSeriesIndexNotValidException: if the index of the specified column is out of bound
        """
        actual_series_number = self.__parameters.get_child_expressions_count()
        if index < 0 or actual_series_number <= index:
            raise UDFInputSeriesIndexNotValidException(index, actual_series_number)
